packages/babel-register/src/worker-client.js

```diff
@@ -0,0 +1,113 @@
+const path = require("path");
+
+const ACTIONS = {
+  GET_DEFAULT_EXTENSIONS: "GET_DEFAULT_EXTENSIONS",
+  SET_OPTIONS: "SET_OPTIONS",
+  TRANSFORM: "TRANSFORM",
+  TRANSFORM_SYNC: "TRANSFORM_SYNC",
+};
+
+class Client {
+  #send;
+
+  constructor(send) {
+    this.#send = send;
+  }
+
+  #eCache;
+  /** @return {string[]} */
+  getDefaultExtensions() {
+    return (this.#eCache ??= this.#send(
+      ACTIONS.GET_DEFAULT_EXTENSIONS,
+      undefined,
+    ));
+  }
+
+  /**
+   * @param {object} options
+   * @return {void}
+   */
+  setOptions(options) {
+    return this.#send(ACTIONS.SET_OPTIONS, options);
+  }
+
+  /**
+   * @param {string} code
+   * @param {string} filename
+   * @return {{ code: string, map: object } | null}
+   */
+  transform(code, filename) {
+    return this.#send(ACTIONS.TRANSFORM, { code, filename });
+  }
+}
+
+// We need to run Babel in a worker because require hooks must
+// run synchronously, but many steps of Babel's config loading
+// (which is done for each file) can be asynchronous
+exports.WorkerClient = class WorkerClient extends Client {
+  // These two require() calls are in deferred so that they are not imported in
+  // older Node.js versions (which don't support workers).
+  // TODO: Hoist them in Babel 8.
+
+  static #worker_threads_cache;
+  /** @type {typeof import("worker_threads")} */
+  static get #worker_threads() {
+    return (WorkerClient.#worker_threads_cache ??= require("worker_threads"));
```

Comment:
Why not just use `return require("worker_threads")`?

Reply:
It throws on older Node.js versions.

Reply:
Uh right, `require` caches internally.